// @sap-ai-sdk/prompt-registry/dist/client/prompt-registry/orchestration-configs-api.js
/*
 * Copyright (c) 2026 SAP SE or an SAP affiliate company. All rights reserved.
 *
 * This is a generated file powered by the SAP Cloud SDK for JavaScript.
 */
import { OpenApiRequestBuilder } from '@sap-ai-sdk/core';
/**
 * Representation of the 'OrchestrationConfigsApi'.
 * This API is part of the 'prompt-registry' service.
 */
export const OrchestrationConfigsApi = {
    _defaultBasePath: undefined,
    /**
     * List orchestration configs
     * @param queryParameters - Object containing the following keys: scenario, name, version, retrieve, include_spec, resolve_template_ref.
     * @param headerParameters - Object containing the following keys: AI-Resource-Group.
     * @returns The request builder, use the `execute()` method to trigger the request.
     */
    listOrchestrationConfigs: (queryParameters, headerParameters) => new OpenApiRequestBuilder('get', '/registry/v2/orchestrationConfigs', {
        queryParameters,
        headerParameters
    }, OrchestrationConfigsApi._defaultBasePath),
    /**
     * Create or update an orchestration config
     * @param body - Request body.
     * @param headerParameters - Object containing the following keys: AI-Resource-Group.
     * @returns The request builder, use the `execute()` method to trigger the request.
     */
    createUpdateOrchestrationConfig: (body, headerParameters) => new OpenApiRequestBuilder('post', '/registry/v2/orchestrationConfigs', {
        body,
        headerParameters
    }, OrchestrationConfigsApi._defaultBasePath),
    /**
     * List orchestration config history
     * @param scenario - Path parameter.
     * @param version - Path parameter.
     * @param name - Path parameter.
     * @param queryParameters - Object containing the following keys: include_spec, resolve_template_ref.
     * @param headerParameters - Object containing the following keys: AI-Resource-Group.
     * @returns The request builder, use the `execute()` method to trigger the request.
     */
    listOrchestrationConfigHistory: (scenario, version, name, queryParameters, headerParameters) => new OpenApiRequestBuilder('get', '/registry/v2/scenarios/{scenario}/orchestrationConfigs/{name}/versions/{version}/history', {
        pathParameters: { scenario, version, name },
        queryParameters,
        headerParameters
    }, OrchestrationConfigsApi._defaultBasePath),
    /**
     * Get orchestration config by UUID
     * @param orchestrationConfigId - Path parameter.
     * @param queryParameters - Object containing the following keys: resolve_template_ref.
     * @param headerParameters - Object containing the following keys: AI-Resource-Group.
     * @returns The request builder, use the `execute()` method to trigger the request.
     */
    getOrchestrationConfigByUuid: (orchestrationConfigId, queryParameters, headerParameters) => new OpenApiRequestBuilder('get', '/registry/v2/orchestrationConfigs/{orchestrationConfigId}', {
        pathParameters: { orchestrationConfigId },
        queryParameters,
        headerParameters
    }, OrchestrationConfigsApi._defaultBasePath),
    /**
     * Delete orchestration config
     * @param orchestrationConfigId - Path parameter.
     * @param headerParameters - Object containing the following keys: AI-Resource-Group.
     * @returns The request builder, use the `execute()` method to trigger the request.
     */
    deleteOrchestrationConfig: (orchestrationConfigId, headerParameters) => new OpenApiRequestBuilder('delete', '/registry/v2/orchestrationConfigs/{orchestrationConfigId}', {
        pathParameters: { orchestrationConfigId },
        headerParameters
    }, OrchestrationConfigsApi._defaultBasePath),
    /**
     * Import orchestration config
     * @param body - Request body.
     * @param headerParameters - Object containing the following keys: AI-Resource-Group.
     * @returns The request builder, use the `execute()` method to trigger the request.
     */
    importOrchestrationConfig: (body, headerParameters) => new OpenApiRequestBuilder('post', '/registry/v2/orchestrationConfigs/import', {
        body,
        headerParameters
    }, OrchestrationConfigsApi._defaultBasePath),
    /**
     * Export orchestration config
     * @param orchestrationConfigId - Path parameter.
     * @param headerParameters - Object containing the following keys: AI-Resource-Group.
     * @returns The request builder, use the `execute()` method to trigger the request.
     */
    exportOrchestrationConfig: (orchestrationConfigId, headerParameters) => new OpenApiRequestBuilder('get', '/registry/v2/orchestrationConfigs/{orchestrationConfigId}/export', {
        pathParameters: { orchestrationConfigId },
        headerParameters
    }, OrchestrationConfigsApi._defaultBasePath)
};
